pkg/bloomcompactor/v2_meta.go

```diff
@@ -0,0 +1,131 @@
+// the ref. Things like index-paths, etc are not needed and possibly harmful
+// in the case we want to do migrations. It's easier to load a block-ref or similar
+// within the context of a specific tenant+period+index path and not couple them.
+type BlockRef struct {
+	OwnershipRange v1.FingerprintBounds
+	Checksum       uint32
+}
+
+func (r BlockRef) Hash(h hash.Hash32) error {
+	if err := r.OwnershipRange.Hash(h); err != nil {
+		return err
+	}
+
+	var enc encoding.Encbuf
+	enc.PutBE32(r.Checksum)
+	_, err := h.Write(enc.Get())
+	return errors.Wrap(err, "writing BlockRef")
+}
+
+type MetaRef struct {
+	OwnershipRange v1.FingerprintBounds
+	Checksum       uint32
+}
+
+// `bloom/<period>/<tenant>/metas/<start_fp>-<end_fp>-<start_ts>-<end_ts>-<checksum>`
+func (m MetaRef) Address(userID string, period int) (string, error) {
+	joined := path.Join(
+		"bloom",
+		fmt.Sprintf("%v", period),
+		userID,
+		"metas",
+		m.OwnershipRange.String(),
+		fmt.Sprintf("%v", m.Checksum),
+	)
+
+	return fmt.Sprintf("%s.json", joined), nil
+}
+
+type Meta struct {
+
+	// The fingerprint range of the block. This is the range _owned_ by the meta and
+	// is greater than or equal to the range of the actual data in the underlying blocks.
+	OwnershipRange v1.FingerprintBounds
+
+	// Old blocks which can be deleted in the future. These should be from pervious compaction rounds.
+	Tombstones []BlockRef
+
+	// The specific TSDB files used to generate the block.
+	Sources []tsdb.SingleTenantTSDBIdentifier
+
+	// A list of blocks that were generated
+	Blocks []BlockRef
+}
+
+// Generate MetaRef from Meta
+func (m Meta) Ref() (MetaRef, error) {
+	checksum, err := m.Checksum()
+	if err != nil {
+		return MetaRef{}, errors.Wrap(err, "getting checksum")
+	}
+	return MetaRef{
+		OwnershipRange: m.OwnershipRange,
+		Checksum:       checksum,
+	}, nil
+}
+
+func (m Meta) Checksum() (uint32, error) {
+	h := v1.Crc32HashPool.Get()
+	defer v1.Crc32HashPool.Put(h)
+
+	_, err := h.Write([]byte(m.OwnershipRange.String()))
+	if err != nil {
+		return 0, errors.Wrap(err, "writing OwnershipRange")
+	}
+
+	for _, tombstone := range m.Tombstones {
+		err = tombstone.Hash(h)
+		if err != nil {
+			return 0, errors.Wrap(err, "writing Tombstones")
+		}
+	}
+
+	for _, source := range m.Sources {
+		err = source.Hash(h)
+		if err != nil {
+			return 0, errors.Wrap(err, "writing Sources")
+		}
+	}
+
+	for _, block := range m.Blocks {
+		err = block.Hash(h)
+		if err != nil {
+			return 0, errors.Wrap(err, "writing Blocks")
+		}
+	}
+
+	return h.Sum32(), nil
+
+}
+
+type TSDBStore interface {
+	ResolveTSDBs() ([]*tsdb.TSDBFile, error)
+}
+
+type MetaStore interface {
+	GetMetas([]MetaRef) ([]Meta, error)
+	PutMeta(Meta) error
+	ResolveMetas(bounds v1.FingerprintBounds) ([]MetaRef, error)
+}
+
+type BlockStore interface {
+	// TODO(owen-d): flesh out|integrate against bloomshipper.Client
+	GetBlocks([]BlockRef) ([]interface{}, error)
```

Comment:
Why do we take and return `interface{}` instead of `v1.Block` here?

Reply:
Mainly because i haven't figured out what the return type should be (which may or may not be block)